_posts/2023-09-07-develop-functions-locally.md

```diff
@@ -0,0 +1,278 @@
+
+```bash
+faas-cli local-run greeter --port 3001
+```
+
+This will run the greeter function and make it available on port `3001`.
+
+```bash
+curl -i http://127.0.0.1:3001
+```
+
+## Run multiple functions
+
+The local-run command is great for running and testing individual OpenFaaS functions but it can only run a single function at a time.
+
+If your stack.yaml file only contains a single function, local-run will run that function by default. When there are multiple functions you need to add the name of the function you want to run as an extra argument to the command.
+
+Create a second function, `echo` and append it to the stack.yml file.
+
+```bash
+faas-cli new greeter --lang node18 --append stack.yml
+faas-cli local-run greeter
+```
+
+Since functions are running as an individual docker container with local-run you can not talk to other functions or the OpenFaaS gateway.
+
+If you are building function pipelines where you need to talk to other functions or if you need to call other services in your cluster, local-run might not be the best option.
+
+While there are some workarounds like port-forwarding the gateway first and making the gateway url configurable in your function through an environment variable you might want to use `faas-cli up --watch` instead.
+
+Running `faas-cli up` will build, push and deploy all functions in the `stack.yml` file. The `--watch` flag will tell the faas-cli to monitor the function source files for any changes and automatically rebuild and redeploy functions as you edit and save your code.
+
+We take a more detailed look into the watch functionality later in this article.
+
+## Use secrets with local-run
+
+All functions can consume secrets in the same way, by reading a file from: `/var/openfaas/secrets/NAME`
+
+To mount a secret in a function the secret name has to be added to the list of secrets in the stack YAML file.
+
+As an example we will add a secret named `api-key` to the echo function:
+
+```yaml
+functions:
+  echo:
+    lang: node18
+    handler: ./echo
+    image: echo:latest 
+    secrets:
+      - api-key
+```
+
+The local-run command looks for secret files in the `.secrets` folder. You will need to create any secrets you want in this location.
+
+All secrets included in a functions stack.yaml will be mounted into the function container so they can be read from their usual location,`/var/openfaas/secrets/NAME, and used within the function.
+
+Create the `.secrets` folder in your local directory and add a file named api-key.
+
+```bash
+mkdir .secrets
+echo "secret-acces-token" > .secrets/api-key
+```
+
+Update the echo function to read the secret file from `/var/openfaas/secrets/api-key` and add the secret value to the response:
+
+```js
+'use strict'
+
+const { readFile } = require('fs').promises
+
+module.exports = async (event, context) => {
+  let token = await readFile("/var/openfaas/secrets/api-key", )
+
+  const result = {
```

Comment:
This seems like a bad idea, returning the secret.

It's not much more effort to turn this into an example that validates an Authorization header, please do that.